api/src/utils/is-a-function.ts

```diff
@@ -0,0 +1,17 @@
+/**
+ * Checks if a string looks like a function call.
+ *
+ * It basically returns true, if the string
+ * -  starts with a word character or an underscore
+ * -  followed by zero or more word characters, digits or underscores
+ * -  followed by an opening parenthesis
+ * -  followed by zero or more characters that are not a closing parenthesis or an opening parenthesis for nested function calls
+ * -  followed by a closing parenthesis
+ *
+ * Spaces are also allowed between the function name and the opening parenthesis, as well as between the closing parenthesis and the arguments.
+ */
+export function valueIsAFunction(value: string): boolean {
+	const regex = /(?:^)([a-zA-Z_][a-zA-Z0-9_]*)\s*\(\s*(?:[^)(]+|\((?:[^)(]+|\([^)(]*\))*\))*\s*\)/g;
```

Comment:
I think it's covered by using this regex, but lets double check and add some tests there's no way to work your way out of a function with a classic bobby tables `some_func(); DROP DATABASE;--`

Reply:
You can probably do `some_func(;DROP DATABASE;)` with this regex?

Reply:
![CleanShot 2024-08-14 at 15 02 40@2x](https://github.com/user-attachments/assets/1db2123e-80b7-468f-8b0d-d3f491cfedfc)

😈

Reply:
Ahhh. Little bobby table. He's grown so big

Reply:
I'm trying to prevent such attacks, but it is difficult

Reply:
i probably start with another approach for now which just allows functions without parameters to reduce the attac vector

Reply:
> i probably start with another approach for now which just allows functions without parameters to reduce the attac vector

I think this would be good enough as a starter 👍 We could even consider only detecting `gen_random_uuid()` as this was the sole function requested in #17697, and would be similar to the current treatment of `now()` etc.

Reply:
done

Reply:
@paescuj true 🤔

Reply:
we introduce something like `uuid()` and use this knex utility: https://knexjs.org/guide/utility.html#uuid. then it would be very similar to `now()`

but I guess this would require changes in the introspection, since `gen_random_uuid()` would probably then to be  mapped to `uuid()`. probably to much to do without having an actual need and considering the circumstances

Reply:
Yeah, I think the current way (or alternatively your suggestion of an explicit allow-list) should do it 👍

Reply:
Having an allow list that is configurable seems like a very good solution, instead of trying to auto-detect function. As there might be users that have an actual string that looks like a function?

Reply:
> Having an allow list that is configurable seems like a very good solution, instead of trying to auto-detect function.

Right, I also tend more towards this ❤️ Have you thought of making it configurable via an environment variable?

> As there might be users that have an actual string that looks like a function?

True, and there might also be the opposite, where a function doesn't necessarily have parentheses like https://www.postgresql.org/docs/current/functions-datetime.html#FUNCTIONS-DATETIME-CURRENT.

Reply:
> Have you thought of making it configurable via an environment variable?

Yeah, that's what I mean by making it configurable :)

Reply:
Might be overkill as discussed, but found another potential solution. We could parse the default value with [`node-sql-parser`](https://www.npmjs.com/package/node-sql-parser) (very interesting package), and obtain the type from the resulting AST:

```js
import nodeSqlParser from 'node-sql-parser';

const parser = new nodeSqlParser.Parser();

function isDbFunction(value) {
  let ast;

  try {
    ast = parser.astify(`SELECT (${value});`);
  } catch {
    return false;
  }
  
  const node = ast[0].columns[0].expr;

  return node.type === 'function';
}

console.log(isDbFunction('test'));                     // -> false
console.log(isDbFunction('1'));                        // -> false
console.log(isDbFunction('CURRENT_DATE'));             // -> true
console.log(isDbFunction('test()'));                   // -> true
console.log(isDbFunction('test("test")'));             // -> true
console.log(isDbFunction('test(); DROP DATABASE;--')); // -> false (parse fail)
console.log(isDbFunction('test(;DROP DATABASE;)'));    // -> false (parse fail)
```

Reply:
that look indeed interesting. But I kept it veery simple now and added a hardcoded whitelist for database functions with the one function from the issue as only entry for now